include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,715 @@
+    void extract_subvector_to (ForwardIterator          indices_begin,
+                               const ForwardIterator    indices_end,
+                               OutputIterator           values_begin) const;
+
+    /**
+     * Read access to the data field specified by @p local_index. The largest
+     * range of the IndexSet is the first one.
+     *
+     * Performance: Direct array access (fast).
+     */
+    Number local_element (const size_type local_index) const;
+
+    /**
+     * Read and write access to the data field specified by @p local_index. The
+     * largest range of the IndexSet is the first one.
+     *
+     * Performance: Direct array access (fast).
+     */
+    Number &local_element (const size_type local_index);
+    //@}
+
+
+    /**
+     * @name 3: Modification of vectors
+     */
+    //@{
+
+    /**
+     * A collective add operation: This function adds a whole set of values
+     * stored in @p values to the vector components specified by @p indices.
+     */
+    template <typename Number2>
+    void add (const std::vector<size_type>  &indices,
+              const std::vector<Number2> &values);
+
+    /**
+     * This is a second collective add operation. As a difference, this
+     * function takes a ReadWriteVector of values.
+     */
+    template <typename Number2>
+    void add (const std::vector<size_type> &indices,
+              const ReadWriteVector<Number2> &values);
+
+    /**
+     * Take an address where <tt>n_elements</tt> are stored contiguously and
+     * add them into the vector. Handles all cases which are not covered by
+     * the other two <tt>add()</tt> functions above.
+     */
+    template <typename Number2>
+    void add (const size_type  n_elements,
+              const size_type *indices,
+              const Number2   *values);
+
+    /**
+     * Prints the vector to the output stream @p out.
+     */
+    void print (std::ostream       &out,
+                const unsigned int  precision  = 3,
+                const bool          scientific = true) const;
+
+    /**
+     * Returns the memory consumption of this class in bytes.
+     */
+    std::size_t memory_consumption () const;
+    //@}
+
+  protected:
+
+    /**
+     * Return the local position of @p global_index.
+     */
+    unsigned int
+    global_to_local (const types::global_dof_index global_index) const
+    {
+      // the following will throw an exception if the global_index is not
+      // in the remaining_elements
+      return static_cast<unsigned int>(stored_elements.index_within_set(global_index));
+    }
+
+    /**
+     * A helper function that is used to resize the val array.
+     */
+    void resize_val (const size_type new_allocated_size);
+
+    /**
+     * Indices of the elements stored.
+     */
+    IndexSet stored_elements;
+
+    /**
+     * Pointer to the array of local elements of this vector.
+     */
+    // TODO: Could we use VectorizedArray here for storage?
+    Number *val;
```

Comment:
@Rombur I think you mean AlignedVector, don't you? And I would suggest to not let std::vector do the memory management because it does initialize the memory on one thread only (if using multithreading, you want something like first-touch memory initialization to have different parts of the vector assigned to different NUMA domains).